### PR TITLE
[qos] process IPv6 packet priority

### PR DESCRIPTION
**Pull request by zhanglongxia**

### CODE REVIEW

File: src/core/net/ip6.cpp [hunk 1]
```diff
@@ -73,6 +73,49 @@ Message *Ip6::NewMessage(uint16_t aReserved, uint8_t aPriority)
         Message::kTypeIp6, sizeof(Header) + sizeof(HopByHopHeader) + sizeof(OptionMpl) + aReserved, aPriority);
 }
 
+uint8_t Ip6::DscpToPriority(uint8_t aDscp)
+{
+    uint8_t priority = Message::kPriorityNormal;
+    uint8_t cs       = aDscp & kDscpCsMask;
+
+    if ((cs == kDscpCs1) || (cs == kDscpCs2))
```
Comment (bukepo) [inline src/core/net/ip6.cpp]: use switch?

File: src/core/thread/mesh_forwarder.cpp [hunk 2]
```diff
@@ -1327,6 +1327,7 @@ void MeshForwarder::HandleFragment(uint8_t *               aFrame,
     Lowpan::FragmentHeader fragmentHeader;
     Message *              message = NULL;
     int                    headerLength;
+    uint8_t                priority;
```
Comment (bukepo) [inline src/core/thread/mesh_forwarder.cpp]: move this declare to line 1339?

File: src/core/net/ip6.cpp (hunk 1 above)
Comment (zhanglongxia) [reply src/core/net/ip6.cpp]: Will update it.

File: src/core/thread/mesh_forwarder.cpp (hunk 2 above)
Comment (zhanglongxia) [reply src/core/thread/mesh_forwarder.cpp]: Will update it.

File: src/core/thread/mesh_forwarder.cpp [hunk 3]
```diff
@@ -1552,6 +1557,46 @@ otError MeshForwarder::HandleDatagram(Message &               aMessage,
...
+    {
+        VerifyOrExit(GetNetif().GetLowpan().DecompressUdpHeader(udpHeader, aFrame, aFrameLength) >= 0);
+    }
+    else
+    {
+        memcpy(&udpHeader, aFrame, sizeof(Ip6::UdpHeader));
+    }
+
+    if (udpHeader.GetDestinationPort() == Mle::kUdpPort ||
+        udpHeader.GetDestinationPort() == OPENTHREAD_CONFIG_JOINER_UDP_PORT ||
```
[the remark this answers is not in the document]
Comment (zhanglongxia) [reply src/core/thread/mesh_forwarder.cpp]: Good question! We should consider the hacking issue. I think it is better to keep the current logic unchanged. I will remove the condition `udpHeader.GetDestinationPort() == OPENTHREAD_CONFIG_JOINER_UDP_PORT`.

**Comment by zhanglongxia** [comment]
@abtink @jwhui Could you give me comments about this PR at you convenience?

File: src/core/net/ip6.cpp [hunk 4]
```diff
@@ -73,6 +73,70 @@ Message *Ip6::NewMessage(uint16_t aReserved, uint8_t aPriority)
...
+    switch (cs)
+    {
+    case kDscpCs1:
+        // fall through
+
+    case kDscpCs2:
+        priority = Message::kPriorityLow;
+        break;
+
+    case kDscpCs0:
```
[the remark this answers is not in the document]
Comment (jwhui) [reply src/core/net/ip6.cpp]: Yes, this follows the [WMM Priority Mapping](https://www.arubanetworks.com/techdocs/Instant_41_Mobile/Advanced/Content/UG_files/Voice%20and%20Video/WMM%20Traffic.htm).

[the remark this answers is not in the document]
Comment (abtink) [reply src/core/net/ip6.cpp]: Got it. Thanks.

File: src/core/net/ip6.cpp [hunk 5]
```diff
@@ -358,6 +422,7 @@ otError Ip6::SendDatagram(Message &aMessage, MessageInfo &aMessageInfo, IpProto
     const NetifUnicastAddress *source;
 
     header.Init();
+    header.SetDscp(PriorityToDscp(aMessage.GetPriority()));
```
Comment (LuDuda) [inline src/core/net/ip6.cpp]: Just a question.. While i am okay with this change, Thread 1.1 specification says that Traffic Class field shall be set to 0 on transmission and ignored on reception - although i 
guess its not a violation. The other downside of this change is that each message with high priority e.g. MLE frame will be increased by 1 byte due to 6LoWPAN compresion. Is that okay?

File: src/core/net/ip6.cpp [hunk 6]
```diff
@@ -73,6 +73,70 @@ Message *Ip6::NewMessage(uint16_t aReserved, uint8_t aPriority)
...
+    case kDscpCs4:
+        // fall through
+
+    case kDscpCs5:
+        // fall through
+
+    case kDscpCs6:
+        // fall through
+
+    case kDscpCs7:
```
[the remark this answers is not in the document]
Comment (zhanglongxia) [reply src/core/net/ip6.cpp]: Will update it.

File: src/core/net/ip6.cpp (hunk 5 above)
Comment (zhanglongxia) [reply src/core/net/ip6.cpp]: The priority of MLE and Thread Mgmt messages are set to kPriorityNet by default. The priority kPriorityNet is mapped to kDscpCs0 in the function `PriorityToDscp()`, the frame won't increase 1 byte due to 6LoWPAN compression. Priority kPriorityNet is identified by the destination UDP port, please refer to the function `GetFramePriority()` in mesh_forwarder.cpp. For user application traffic, in order to get better performance, I think it is worth adding a byte.

**Comment by zhanglongxia** [comment]
@abtink @LuDuda Could you review this PR again at your convenience.

**Pull request merged**